staging/src/k8s.io/kubectl/pkg/cmd/apply/applyset_pruner.go

```diff
@@ -16,15 +16,114 @@ limitations under the License.
 
 package apply
 
-import "fmt"
+import (
+	"context"
+	"fmt"
+	"io"
+
+	corev1 "k8s.io/api/core/v1"
+	"k8s.io/apimachinery/pkg/api/meta"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/util/sets"
+	"k8s.io/cli-runtime/pkg/printers"
+	"k8s.io/client-go/dynamic"
+	cmdutil "k8s.io/kubectl/pkg/cmd/util"
+)
 
 type applySetPruner struct {
+	dynamicClient dynamic.Interface
+
+	visitedUids sets.String
+
+	cascadingStrategy metav1.DeletionPropagation
+	dryRunStrategy    cmdutil.DryRunStrategy
+	gracePeriod       int
+
+	toPrinter func(string) (printers.ResourcePrinter, error)
+
+	out io.Writer
+}
+
+func newApplySetPruner(o *ApplyOptions) *applySetPruner {
+	return &applySetPruner{
+		dynamicClient: o.DynamicClient,
+
+		visitedUids: o.VisitedUids,
+
+		cascadingStrategy: o.DeleteOptions.CascadingStrategy,
+		dryRunStrategy:    o.DryRunStrategy,
+		gracePeriod:       o.DeleteOptions.GracePeriod,
+
+		toPrinter: o.ToPrinter,
+
+		out: o.Out,
+	}
 }
 
-func newApplySetPruner(_ *ApplyOptions) *applySetPruner {
-	return &applySetPruner{}
+func (p *applySetPruner) pruneAll(ctx context.Context, applyset *ApplySet) error {
+	applysetLabelSelector := applyset.LabelSelectorForMembers()
+
+	opt := metav1.ListOptions{
+		LabelSelector: applysetLabelSelector,
+	}
+
+	// TODO: Split into discovery and deletion, run discovery in parallel (and maybe in consistent order or in parallel?)
+	for _, restMapping := range applyset.AllPrunableResources() {
+		switch restMapping.Scope.Name() {
+		case meta.RESTScopeNameNamespace:
+			for _, namespace := range applyset.AllPrunableNamespaces() {
+				if err := p.prune(ctx, namespace, restMapping, opt); err != nil {
+					return fmt.Errorf("error pruning namespaced object %v: %w", restMapping.GroupVersionKind, err)
+				}
+			}
+
+		case meta.RESTScopeNameRoot:
+			if err := p.prune(ctx, metav1.NamespaceNone, restMapping, opt); err != nil {
+				return fmt.Errorf("error pruning nonNamespaced object %v: %w", restMapping.GroupVersionKind, err)
+			}
+
+		default:
+			return fmt.Errorf("unhandled scope %q", restMapping.Scope.Name())
+		}
+	}
+
+	return nil
+}
+
+func (p *applySetPruner) prune(ctx context.Context, namespace string, mapping *meta.RESTMapping, opt metav1.ListOptions) error {
+	objects, err := p.dynamicClient.Resource(mapping.Resource).Namespace(namespace).List(ctx, opt)
+	if err != nil {
+		return err
+	}
+
+	for i := range objects.Items {
+		obj := &objects.Items[i]
+
+		annotations := obj.GetAnnotations()
+		if _, ok := annotations[corev1.LastAppliedConfigAnnotation]; !ok {
```

Comment:
This should be checking our membership label, and should definitely not depend on the CSA annotation. In fact, we should have tests with both CSA and SSA to prove that.

Reply:
Good call.  This code mirrors the old applyv1 logic, but we should make it v2 :-)  

I am starting by replacing the annotations check with a comment.

We are actually applying our label selector in the list (via opt), but I am going to reshuffle the arguments to make it obvious that we are doing that.